SimTracker/TrackTriggerAssociation/plugins/TTTrackAssociator.cc

```diff
@@ -0,0 +1,203 @@
+            {
+              auxMap.insert( std::make_pair( testTP.get(), testTP ) );
+            }
+
+          }
+        } /// End of loop over the clusters
+
+        /// Check if the stub is unknown
+	if ( TTStubAssociationMapHandle->isUnknown( theseStubs.at(is) ) )
+	  ++mayCombinUnknown;
+
+      } /// End of loop over the stubs
+
+      /// If there more than 2 unknown stubs, go to the next track
+      /// as this track may be COMBINATORIC or UNKNOWN
+      if ( mayCombinUnknown >= 2 )
+      	continue;
+
+      /// If we are here, all the stubs are either combinatoric or genuine
+      /// and there is no more than one fake stub in the track
+      /// Loop over all the TrackingParticle which have been found in the track at some point
+      /// (stored in auxMap)
+
+      std::vector< const TrackingParticle* > tpInAllStubs;
+
+      std::map< const TrackingParticle*, edm::Ptr< TrackingParticle > >::const_iterator iterAuxMap;
+      for ( iterAuxMap = auxMap.begin();
+            iterAuxMap != auxMap.end();
+            ++iterAuxMap )
+      {
+        /// Get all the stubs from this TrackingParticle
+        std::vector< edm::Ref< edmNew::DetSetVector< TTStub< Ref_Phase2TrackerDigi_ > >, TTStub< Ref_Phase2TrackerDigi_ > > > tempStubs = TTStubAssociationMapHandle->findTTStubRefs( iterAuxMap->second );
+
+	int nnotfound=0;
+        //bool allFound = true;
+        /// Loop over the stubs
+	//        for ( unsigned int js = 0; js < theseStubs.size() && allFound; js++ )
+        for ( unsigned int js = 0; js < theseStubs.size(); js++ )
+        {
+          /// We want that all the stubs of the track are included in the container of
+          /// all the stubs produced by this particular TrackingParticle which we
+          /// already know is one of the TrackingParticles that released hits
+          /// in this track we are evaluating right now
+          if ( std::find( tempStubs.begin(), tempStubs.end(), theseStubs.at(js) ) == tempStubs.end() )
+          {
+	    //  allFound = false;
+	    ++nnotfound;
+          }
+        }
+
+        /// If the TrackingParticle does not appear in all stubs but one
+        /// then go to the next track
+        if (nnotfound>1)
+	  //if (!allFound)
+	  continue;
+
+        /// If we are here, it means that the TrackingParticle
+        /// generates hits in all stubs but one of the current track
+        /// so put it into the vector
+        tpInAllStubs.push_back( iterAuxMap->first );
+      }
+
+      /// Count how many TrackingParticles we do have
+      std::sort( tpInAllStubs.begin(), tpInAllStubs.end() );
+      tpInAllStubs.erase( std::unique( tpInAllStubs.begin(), tpInAllStubs.end() ), tpInAllStubs.end() );
+      unsigned int nTPs = tpInAllStubs.size();
+
+      /// If only one TrackingParticle, GENUINE
+      /// if different than one, COMBINATORIC
+      if ( nTPs != 1 )
+      {
+        continue;
+	if (nTPs>1) std::cout << "This should never happen!!!" << std::endl;
```

Comment:
please use log warning (but ok, this line can never be executed - maybe it can be removed?